Added autocomplete to the 'Redirect To: field

bmarshall511 opened the pull request:
### Description of the Change

Added a styled autocomplete to the 'Redirect To:' field to make it easier for users to enter URLs. Married the styles of WP's default admin settings page & Gutenberg's autocomplete link popover component.

Closes #41

### How to test the Change

1. With existing posts/pages in place, go to /wp-admin/post-new.php?post_type=redirect_rule
2. In the 'Redirect To:' field, beginning typing to show a list of matched posts

### Changelog Entry

> Added - Added autocomplete to the 'Redirect To:' field


### Credits

Props @bmarshall511 


### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).

bmarshall511 committed:
feat(autocomplete): added autocomplete to the redirect to field

ravinderk commented on inc/classes/class-srm-post-type.php:
```diff
@@ -671,18 +673,65 @@ public static function factory() {
+		$search_term = ltrim( $search_term, '/' );
+
+		$query = get_posts(
+			array(
```
@bmarshall511 Is there any reason we limit search results to post type `post`?

bmarshall511 replied:
Good catch! The limitation to the `post` post type was indeed an oversight. I've updated the code to search across all post types.

ravinderk commented on inc/classes/class-srm-post-type.php:
```diff
@@ -671,18 +673,65 @@ public static function factory() {
+	 * @return void
+	 */
+	public function srm_autocomplete() {
+		check_ajax_referer( 'srm_autocomplete_nonce', 'security' );
```
@bmarshall511 Can we also check user capability, whether the user can read/search other post types?

bmarshall511 replied:
Definitely! I've updated the code to include a check for the `read` capability.

ravinderk replied:
@bmarshall511 Can we use [`srm_manage_redirects`](https://github.com/10up/safe-redirect-manager/blob/832ca3411801a4768772410d8dc8284457f9600d/inc/classes/class-srm-post-type.php#L469)?

bmarshall511 replied:
Sure! Updated accordingly.

ravinderk replied:
@bmarshall511 pardon me 🙏 . My last comment was for something other than renaming the nonce action name. [srm_manage_redirects](https://github.com/10up/safe-redirect-manager/blob/832ca3411801a4768772410d8dc8284457f9600d/inc/classes/class-srm-post-type.php#L469) is a capability and we can check for this capability instead of `read`. 

Let me know if you have questions.

bmarshall511 replied:
Whoops! Yeah, my misunderstanding. I've rectified it. 🤦‍♂️

ravinderk commented on inc/classes/class-srm-post-type.php:
```diff
@@ -49,6 +49,8 @@ public function setup() {
 		add_filter( 'default_hidden_columns', array( $this, 'filter_hidden_columns' ), 10, 1 );
 		add_action( 'admin_enqueue_scripts', array( $this, 'load_resources' ), 10, 0 );
 		add_action( 'wp_ajax_srm_validate_from_url', array( $this, 'srm_validate_from_url' ), 10, 0 );
+		add_action( 'wp_ajax_nopriv_srm_autocomplete', array( $this, 'srm_autocomplete' ), 10, 0 );
```
@bmarshall511 Is there any reason we are handling Ajax requests from unauthenticated users?

bmarshall511 replied:
Thanks for pointing this out! Another oversight on my part. I'm accustomed to writing Ajax functions for the frontend, didn't take into account that this is strictly backend functionality. The code has been updated accordingly.

ravinderk reviewed (changes requested):
@bmarshall511 I left a few questions on the pull request.

bmarshall511 committed:
fix(qa): minor code updates based on code review

bmarshall511 commented:
@ravinderk Thanks for the feedback! I've addressed the points you raised & updated the code accordingly. Let me know if there's anything else I missed.

bmarshall511 committed:
fix(qa): updated the capability for the ajax call
